[Runtime] Add __extendhfxf2 support to the Swift runtime.

al45tair opened the pull request:
It turns out that we can just use the `Float16` to `Float32` conversion and let the compiler turn that into a `long double` for us, which means we can take advantage of the F16C CPU instructions if they're present as well.

rdar://104134160

al45tair committed:
[Runtime] Add __extendhfxf2 support to the Swift runtime.

It turns out that we can just use the Float16 to Float32 conversion and
let the compiler turn that into a long double for us, which means we
can take advantage of the F16C CPU instructions if they're present as well.

rdar://104134160

al45tair commented:
@swift-ci Please smoke test

compnerd commented on stdlib/public/runtime/Float16Support.cpp:
```diff
@@ -150,6 +151,14 @@ SWIFT_RUNTIME_EXPORT unsigned short __truncdfhf2(double d) {
+/// of Float16 values and since we already know how to convert Float32 to long
+/// double (which, at least on x86, doesn't involve function calls), we just
+/// let the compiler do the latter part for us.
+SWIFT_RUNTIME_EXPORT long double __extendhfxf2(short h) {
+  return __gnu_h2f_ieee(h);
+}
```
We should add this conditionally I think.  FP80 is not supported on Windows, Android, and Apple platforms.

```suggestion
#if (defined(__i386__) || defined(__x86_64__)) && !(defined(__ANDROID__) || defined(__APPLE__) || defined(_WIN32))

SWIFT_RUNTIME_EXPORT long double __extendhfxf2(short h) {
  return __gnu_h2f_ieee(h);
}

#endif
```

al45tair replied:
Do we think conditionalising it is really necessary? `long double` compiles everywhere these days, no? (On platforms without native support for anything larger, it tends to turn into `double`, possibly stored in a larger type, I think? On PowerPC I think it would have been that weird thing with two `double` values in it.)

tbkka replied:
This clearly does the right thing for any "long double", so there's no harm in compiling it on any platform, although because it has `xf` in the name, it should only be called on platforms where `long double` is in fact Intel 80-bit extended format.

compnerd replied:
Yes, but it does add to the runtime (and standard library size).  Plus, it is an extra interface that now becomes part of the ABI of the library.

al45tair replied:
Additionally (and I hadn't paid enough attention to this), there is already a huge conditional wrapping the entire file, which will need adjusting slightly as some of these functions will be needed for Linux.

al45tair commented:
@swift-ci Please smoke test

compnerd commented:
I think that we shouldn't add the additional functions and should prefer to bump the baseline.  These functions will be ABI and we will be unable to remove them.  For platforms (e.g. Darwin) where we are using dynamic linking for the builtins (libSystem re-exports libcompiler_rt.builtins-*.dylib), this does technically increase the surface of the ODR violations.

stephentyrone commented:
The functions will be ABI _if we tag a release before removing them_.

compnerd commented:
@stephentyrone true, but we do not have a good way to ensure that we do not forget about this currently.

stephentyrone commented:
@swift-ci smoke test

tbkka commented:
Windows/macOS test results are no longer available, but I got this from the Linux results:
```
FixedPointConversion_Debug64_ToInt-2849e2.o:function main:
error: undefined reference to '__extendhfsf2'
```

tbkka commented on stdlib/public/runtime/Float16Support.cpp:
```diff
@@ -150,6 +151,29 @@ SWIFT_RUNTIME_EXPORT unsigned short __truncdfhf2(double d) {
+//
+// There's no risk of rounding problems from the double conversion, because
+// we're extending.
+SWIFT_RUNTIME_EXPORT long double __extendhfxf2(short h) {
```
If I'm reading the conditionals correctly, this entire file is compiled only for
* Android on ARM
* Windows on x86
If we need this function on macOS or Linux, we'll need to shuffle things around a bit.

tbkka replied:
(See Lines 32-34 above)

al45tair replied:
We don't need it on macOS. I think we will need it on Linux. I'd been concentrating on getting it right on Windows because it was failing for other reasons there. I'm looking at Linux now — I suspect it's just a matter of getting the functions inside the correct conditional.

tbkka commented:
@swift-ci Please test Windows platform

tbkka commented:
Re-running Windows tests to try to figure out why Alastair's fix here isn't working there.

tbkka commented:
@swift-ci Please test Windows

al45tair commented:
The Windows tests fail because the ABI changed under us (those functions used to take F16 arguments in `rdi`, but as of LLVM 15 they take them in `xmm0` instead, it seems). Looking at how to fix that now.

al45tair committed:
[Runtime] Adjust argument types to match the LLVM15 ABI.

It appears that LLVM 15 changed the ABI for _Float16 on x86-64 such that
values are now passed in `xmm0` instead of using integer registers.

Also enable this code for Linux.

rdar://104134160

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test
